Implement forward AD for linalg.svd and improve svd_backward

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70253
* #69827

I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D33751982](https://our.internmc.facebook.com/intern/diff/D33751982)

lezcano committed:
Implement forward AD for linalg.svd and improve svd_backward

I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+  return std::make_tuple(std::move(dU), std::move(dS), std::move(dVh));
+}
+
+// This makes no assumption on the signs of sigma.
```
sigma is the S matrix here right?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+                         : at::matmul(at::matmul(U.mH(), dA), V);
+
+  // We don't want dS to be a view into a larger tensor so we clone it
+  auto dS = is_complex ? at::real(dP.diagonal(0, -2, -1)).clone()
```
Why do you clone here? It is fine for the gradient to be a view of the input.

Lezcano replied:
For two reasons:
1. We don't want `dS` (a vector) to be a view into a full matrix that otherwise could be released from memory
2. I'm setting the diagonal of `dP`  to zero in the next line, so `dS` cannot be just a view.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+  auto dS = is_complex ? at::real(dP.diagonal(0, -2, -1)).clone()
+                       : dP.diagonal(0, -2, -1).clone();
+
+  // dP = dP - dS
```
Not sure what this comment means?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+
+  // Im(diag(dP)) / (2S)
+  auto ImdiagdP2S = is_complex ? at::imag(dP.diagonal(0, -2, -1)).div(2. * S)
+                               : Tensor{};
```
You can do `Tensor()` as well.

Lezcano replied:
I like the uniform initialisation syntax for empty constructors, but yeah.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+  // that just depends on U^H gU.
+
+  // Trivial case
+  if (!gS.defined() && !gU.defined() && !gVh.defined()) {
```
This case should never happen?
But we can keep it

Lezcano replied:
Oh, I did not know that this case never happens. We start many linalg backward functions like this. I'll remove it. cc @nikitaved

Lezcano replied:
Actually, I believe we cannot remove it as per the last item in
https://github.com/pytorch/pytorch/blob/9266b2af73311792a746a2f741cc13145b3701df/tools/autograd/derivatives.yaml#L151-L169

Lezcano replied:
nevermind, point 2 applies and we don't need this here :D

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+                                           : at::zeros_like(S);
+    // Rather lax atol and rtol, as we don't want false positives
+    TORCH_CHECK(at::allclose(imdiag_UhgU, -imdiag_VhgV, /*rtol=*/1e-2, /*atol=*/1e-2),
+                "The singular vectors in the complex case are specified up to multiplication "
```
We should mention that this is raised from within svd backward no?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2282,103 +2282,323 @@ Tensor slice_backward_wrapper(
+      const auto E = [&S]{
+        const auto S2 = S * S;
+        auto ret = S2.unsqueeze(-2) - S2.unsqueeze(-1);
+        // Any number a != 0 would, as we are going to compute 0 / a
```
Part of the sentence is missing no?

Lezcano replied:
I'll write
>Any number a != 0 would, as we are just going to use it to compute 0 / a later on

Lezcano commented on tools/autograd/derivatives.yaml:
```diff
@@ -1457,7 +1457,15 @@
 
 - name: linalg_svd(Tensor A, bool full_matrices=True) -> (Tensor U, Tensor S, Tensor Vh)
-  A: svd_backward(grads, full_matrices, U, S, Vh)
+  A: "svd_backward(full_matrices && grad_U.defined() ? grad_U.narrow(-1, 0, S.size(-1)) : grad_U,
```
Just to have the preprocessing already done within the function. Even though `svd_backward` is called from other places, this is the only place from which you can set `full_matrices=True`, so if we keep this logic outside, we don't need to pass any `full_matrices=False` from the other places (or even better, a user cannot call this function with `full_matrices=True`).

albanD replied:
Sounds good

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

Lezcano commented:
@pytorchbot ciflow rerun

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

IvanYashchuk commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+    return ret;
+  }();
+
+  const auto sym = [](const Tensor& X) { return X + X.mH(); };
```
Usually symmetrization is done with `X -> (X + X.mH()) / 2` so that if X is already a Hermitian matrix it's not changed. Is omitting 1/2 factor done on purpose here?
Tests pass so everything is good then.

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

CI failures are real though.

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

albanD commented:
The rocm failures are relevant right?

IvanYashchuk commented:
Yes, the ROCm failures are relevant. They come from https://github.com/pytorch/pytorch/pull/69827 in the stack.

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33751982).

nikitaved commented:
@mruberry , I would also like to have a look at this PR if you do not mind.

mruberry commented:
> @mruberry , I would also like to have a look at this PR if you do not mind.

Sounds great! Ping me when you're done. Note there is an internal failure on the PR preceding this one in the stack that's also preventing landing this.

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

Differential Revision: [D33751982](https://our.internmc.facebook.com/intern/diff/D33751982)

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -285,6 +285,9 @@ Tensor linalg_pinv(
     const optional<Tensor>& atol_opt,
     const optional<Tensor>& rtol_opt,
     bool hermitian) {
+  // FIXME: Whenever we have a nice lstsq, we should dispatch this function to simply be
+  // `torch.lstsq(A, torch.eye(A.shape[-1]), atol=atol, rtol=rtol)`
+  // with a driver that supports singular inputs
```
gelsy and all SVD-based drivers do support singular inputs but only for the CPU.

IvanYashchuk replied:
For CUDA only `driver='gels'` is implemented.

Lezcano replied:
What I mean there is that it'd be good to move the code we have here into `lstsq` to support "manually" the SVD modes on GPU as well. This way we could have this functionality here and in `lstsq`.

nikitaved replied:
We can still dispatch it for the CPU if needed.

nikitaved commented on tools/autograd/derivatives.yaml:
```diff
@@ -1467,7 +1467,15 @@
-  A: svd_backward(grads, full_matrices, U, S, Vh)
+  A: "svd_backward(full_matrices && grad_U.defined() ? grad_U.narrow(-1, 0, S.size(-1)) : grad_U,
+                   grad_S,
+                   full_matrices && grad_Vh.defined() ? grad_Vh.narrow(-2, 0, S.size(-1)) : grad_Vh,
+                   full_matrices ? U.narrow(-1, 0, S.size(-1)) : U,
+                   S,
+                   full_matrices ? Vh.narrow(-2, 0, S.size(-1)) : Vh)"
```
the narrowing is no longer done in the backward?

Lezcano replied:
I moved it here to remove the `full_matrices` keyword from the backwards. `svd_backwards` is used in a number of places, and I wanted to make sure no one could call it with `full_matrices=True`, as the usual caveat applies: svd with `full_matrices=True` is not a well-defined function so, in particular, it is not differentiable.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+  const auto n = dA.size(-1);
+  const auto k = S.size(-1);
+
+  const auto U = full_matrices ? U_.narrow(-1, 0, k) : U_;
+  const auto Vh = full_matrices ? Vh_.narrow(-2, 0, k) : Vh_;
```
You did this narrowing in the derivatives.yaml for backward and not for the jvp. I'd rather have this narrowing done in the backward/forward AD code for simplicity. There are some methods that depend on the older svd_backward which handled narrowing.

Lezcano replied:
Ugh, I forgot to remove this code.

I made sure all the methods that call into `svd_backward` do not need to prepare anything. The rationale is: https://github.com/pytorch/pytorch/pull/70253/files/3c438217696f4403766bdfadc466a05c46961a69#r791727546

Lezcano replied:
Nevermind, I kept it here because I need to know the size of `U` and `Vh` to return tensors of the correct size. Note that, again, the filling of those tensors is arbitrary, and this function would not pass autograd by itself.
I'm not 100% sure why we even support it, and even less as the default option. Cf. the discussion in https://github.com/data-apis/array-api/issues/214
cc  @albanD for reference

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+  //
+  // To think about M, consider the case case k = 1. The, we have the bundle
+  // pi : St_1(C^n) x U(1) -> M
+  // now, St_1(C^n) are just vectors of norm 1 in C^n. That's exactly S^{2n-1}.
```
What is `S^{2n - 1}` and why is `{2n-1}` in there?

Lezcano replied:
The sphere of 2n-1 dimensions. It is just the complex vectors of norm 1, since C^n \iso R^{2n}, so the complex vectors of norm 1 is exactly S^{2n-1}. You can define it as S^{2n-1} = { z \in C^n | z^H z = 1}

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+  // pi : St_1(C^n) x U(1) -> M
+  // now, St_1(C^n) are just vectors of norm 1 in C^n. That's exactly S^{2n-1}.
+  // Then, in this case, we're quotienting out U(1) completely, so we get that
+  // pi : S^{2n-1} x U(1) -> CP(n-1)
```
What is `CP`?

Lezcano replied:
The [complex projective space](https://en.wikipedia.org/wiki/Complex_projective_space) of dimension n-1.

Lezcano replied:
I'll add these two points of notation to the note though. Thanks!

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+  // Using this in the derviaton for the forward AD, one sees that, with the notation from those notes
+  // Using this and writing sym(X) = X + X^H, we get that the forward AD for SVD in the complex
+  // case is given by
+  // dU = U (sym(dX S) / E + i Im(diag(dX)) / (2S))
```
was `dX` defined above?

Lezcano replied:
It is defined below, together with all the other quantities.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+  // Denote by skew(X) = X - X^T, and by A o B the coordinatewise product, then
+  // if m == n
+  //   gA = U [(F o skew(U^T gU))S + S(F o skew(V^T gV)) + I o gS ]V^T
+  // where F_jk = 1 / (S_j - S_k) if j != k and zero otherwise
```
I think it is `S_k - S_j`.

Lezcano replied:
It should be `1/(S_k^2 - S_j^2)`, but yeah. I corrected all the instances of this mistake, and I always referred to this matrix as E and divided by it, as it's done in the implementation.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+    const auto imdiag_VhgV = gVh.defined() ? at::imag(VhgV.diagonal(0, -2, -1))
+                                           : at::zeros_like(S);
+    // Rather lax atol and rtol, as we don't want false positives
+    TORCH_CHECK(at::allclose(imdiag_UhgU, -imdiag_VhgV, /*rtol=*/1e-2, /*atol=*/1e-2),
+                "svd_backward: The singular vectors in the complex case are specified up to multiplication "
+                "by e^{i phi}. The specified loss function depends on this phase term, making "
+                "it ill-defined.");
+  }
```
This will cause syncing, right? I guess we should put it into the doc.

Lezcano replied:
It does sync. And yup, I'll add it to the doc.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
+                "it ill-defined.");
+  }
+
+  // gA = (skew(U^H gU) / E) S +  S ((skew(V^H gV) / E) + I o (gS + i Im(diag(U^H gU)) / S)
+  Tensor gA = [&] {
+    // ret holds everything but the diagonal of gA
```
In some parts `E` is referred to as `F`, so a bit confusing.

Lezcano replied:
So, in all these functions (SVD / eig / eigh), when you divide a matrix by `E`, the diagonal of that matrix is already zero. As such, I define the diagonal of `E` in all these functions to be `1`, but any value other than `0` would do.

nikitaved replied:
But division by 1 and multiplication with 0 is not the same.

nikitaved replied:
Ah, sorry, you have 0 / E, so all good.

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

Differential Revision: [D33751982](https://our.internmc.facebook.com/intern/diff/D33751982)

[ghstack-poisoned]

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,325 @@ Tensor slice_backward_wrapper(
+      }();
+
+      if (gU.defined()) {
+        if (gVh.defined()) {
+          return (skew(UhgU) / E) * S.unsqueeze(-2) + S.unsqueeze(-1) * (skew(VhgV) / E);
+        } else {
```
maybe we can delay `/ E` to the very end?

Lezcano replied:
Now that you mention it... I'm pretty sure I thought about this and I thought then that you can't, but I don't see why now. Let me try it.

Lezcano replied:
It seems to work! Changed. Thanks!

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,325 @@ Tensor slice_backward_wrapper(
+    return at::matmul(dVhaux, Vh);
+  }();
+  if (m < n) {
+    // dVh += S^{-1} U^H dA (I_n - VV^H)
+    const auto UHdASinv = at::matmul(U.mH() / S.unsqueeze(-1), dA) ;
+    dVh += UHdASinv - at::matmul(at::matmul(UHdASinv, V), Vh);
```
perf nit: maybe we can write a function that makes these products that reduce over the largest dimension and do not produce `max(m, n) x max(m, n)` matrices.

nikitaved replied:
In fact, we can write a utility function that performs these kind of projections, as this one is quite heavily used in the backward context.

Lezcano replied:
So, if I'm not mistaken, this never instantiates a large matrix. Note that `V` is of size `(n, m)` with `n > m`. As such, doing `UHdASinv @ V` gives a small matrix of size `(n, n)`. Then this matrix is mapped back onto the large space by multiplying by `Vh`. The size of the largest matrix is the size of the input `dA`.

I tried to be careful in this sense when writing all this code. Please tell me if you find any place where this does not hold, but I believe it holds here.

nikitaved replied:
Exactly, `n > m`, but a matrix `n x n` is produced.

Lezcano replied:
When? I don't think so...

Lezcano replied:
sorry, it should said `(m, m)` over there. Editing

nikitaved replied:
It just follow from your comment.

Lezcano replied:
No `(n, n)` matirx is computed here. This is the correct way of associating things when dealing with a projection. Note that `V` is a tall matrix and `A` is a wide matrix, so the result is a small matrix

nikitaved replied:
All good, sorry, I forgot that U is `(m, m)`.

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

Differential Revision: [D33751982](https://our.internmc.facebook.com/intern/diff/D33751982)

[ghstack-poisoned]

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,325 @@ Tensor slice_backward_wrapper(
+                    const Tensor& U,
+                    const Tensor& S,
+                    const Tensor& Vh) {
+  // Throughout both the real and complex case we assume A has distinct singular values.
+  // Furthermore, if A is rectangular or complex, we assume it's full-rank.
+  //
```
Having rank `min(m, n) - 1` does not work?

Lezcano replied:
Not when A is rectangular, as you need to compute `S^{-1}`

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2357,103 +2357,325 @@ Tensor slice_backward_wrapper(
+    return ret;
+  }();
+
+  if (m > n && gU.defined()) {
+    // gA = [UgA + (I_m - UU^H)gU S^{-1}]V^H
+    gA  = at::matmul(U, gA);
+    const auto gUSinv = gU / S.unsqueeze(-2);
+    gA = gA + gUSinv - at::matmul(U, at::matmul(U.mH(), gUSinv));
```
Just to avoid any confusion. Do I understand that correctly that rank of U and V is at most `min(m, n)` due to the narrowing?

Lezcano replied:
If by rank you mean the dimension of the image, then, since `U` and `V` have orthogonal columns they are always full rank.

nikitaved replied:
No, I mean the number of vectors.

Lezcano replied:
If `m > n`, `U` is of size `m x n` and `V` of `n x n`.
If `n > m`, `U` is of size `m x m`  and `V` is of size `n x m`.
I mean, the usual stuff in a compact SVD, one of the two matrices (if you consider V^H) is the size of A and the other one is square and small.

nikitaved replied:
So, exactly as I thought, at most `min(m, n)`. Because of `full_matrices=True` by default I thought that V and U are the bases and hence had this confusion with probably inefficient matrix product associations. I wonder why would we need `full_matrices=True` by default anyway?

Lezcano replied:
That default is absolute nonsense. I tried to fight to get it changed in https://github.com/data-apis/array-api/issues/214 but you can see how the discussion went...

nikitaved replied:
I believe having `full_matrices=True` is not useful, and, moreover, it inflicts an unnecessary memory penalty for super wide/tall matrices.

nikitaved replied:
I see, we have to carry all those bases for super tall/wide matrices and suffer the consequences of low-memory :)

Lezcano replied:
Note that these matrices are trimmed in `derivatives.yaml` before calling this funciton if `full_matrices=True`, so it's fine. The main problem is that, since it's the default, if a user is unwary, doing `u, s, vh = torch.linalg.svd(A)` will give them unnecessarily large matrices.

nikitaved replied:
But they are trimmed with a narrow, right? This one is a view, so I am not sure it saves us...

Lezcano replied:
We could use a `clone()`, but I think it's better just to optimise for the case where `full_matrices=False`, as it's the one most people will use.

nikitaved replied:
Nothing we can do, if it is computed by the forward, the memory is already allocated so we might just use a view in the backward :)

nikitaved commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5547,29 +5536,41 @@ def sample_inputs_svd(op_info, device, dtype, requires_grad=False, **kwargs):
-    # See the docs of torch.linalg.svd for more info
-    def check_grads(usv):
+    def uniformize(usv):
         S = usv[1]
         k = S.shape[-1]
         U = usv[0][..., :k]
         Vh = usv[2] if is_linalg_svd else usv[2].mH
         Vh = Vh[..., :k, :]
-        return (U.abs(), S, Vh.abs())
+        return U, S, Vh
```
Do you have a test which checks the error message when the objective function does not satisfy the invariant?

Lezcano replied:
I do not, as these tests are rather brittle, as the errors might change and so on. Also, testing an error in the backward within current OpInfos might be tricky.

nikitaved replied:
We can do so in `test_linalg.py`. Looks like just summing up the inputs will break the invariance, right?

Lezcano replied:
Sure, I'll add a test.

nikitaved commented:
I have a general question. Do we want to keep the check for the invariance in the backward? That seems like a significant perf penatly. Maybe it is better to mention this invariance in the documentation? Granted that checking it from the user interface is problematic, but then the complex case is quite non-trivial so maybe the user is already aware of potential issues...

Lezcano commented:
I chose this design because, as you said, this function is already rather difficult, so whatever we can do to improve the UX will be welcome by them. Note that in JAX (and by extension TF), they wanted to do this, but they did not know how to do it. See this comment https://github.com/google/jax/issues/2748#issuecomment-644880044 in particular and the rest of that issue / linked issues for context.

Also, it is true that this synchronises but:
- The SVD is already a very heavy function, so well.
- The check just happens in the complex case, which is not used that much.
- We already do this in `eig` and `eigh`:
https://github.com/pytorch/pytorch/blob/40e88b75c423ef2e70ff3fce9d7ffcafea05872b/torch/csrc/autograd/FunctionsManual.cpp#L2644-L2651

lezcano committed:
Update on "Implement forward AD for linalg.svd and improve svd_backward"


I included a derivation of the formula in the complex case, as it is
particularly tricky. As far as I know, this is the first time this formula
is derived in the literature.

I also implemented a more efficient and more accurate version of svd_backward.
More importantly, I also added a lax check in the complex case making sure the loss
function just depends on the subspaces spanned by the pairs of singular
vectors, and not their joint phase.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

Differential Revision: [D33751982](https://our.internmc.facebook.com/intern/diff/D33751982)

[ghstack-poisoned]

nikitaved commented on test/test_linalg.py:
```diff
@@ -2817,6 +2817,27 @@ def test_svd(self, device, dtype):
                 self.assertEqual(S_s, S)
 
+    @skipCUDAIfNoMagmaAndNoCusolver
+    @skipCPUIfNoLapack
+    @dtypes(torch.complex128)
+    def test_invariance_error_spectral_decompositions(self, device, dtype):
+        make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=True)
+        A = make_arg((3, 3))
+        with self.assertRaisesRegex(RuntimeError, "ill-defined"):
+            U, _, Vh = torch.linalg.svd(A, full_matrices=False)
+            (U + Vh).sum().backward()
```
That reminds me of the condition check. Shall we have a float/double-dependent atol/rtol?

Lezcano replied:
I think it's good as it is. I wanted to put something that won't fire false positives, (hence the very lax condition in the `all_close`). I think the tolerances I put should be good. We just want to catch any violation of the invariance, which should give a tensor that's manifestly non-zero when the condition's tested.

nikitaved replied:
I see these thresholds are used for all the spectral decompositions. Seems quite arbitrary but all right...

nikitaved reviewed (approved):
LGTM! Thanks you, Mario! Your call, @mruberry !

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33751982).

Lezcano commented:
Failures seem unrelated

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2732,7 +2732,7 @@ Tensor linalg_cond(const Tensor& self, const optional<Scalar>& opt_ord) {
   // If ord == None or ord == ±2
   if (std::abs(ord.toDouble()) == 2.0) {
-    auto singular_values = std::get<1>(at::svd(self));
+    auto singular_values = at::linalg_svdvals(self);
```
This diff is most likely the reason for the XLA tests failure. https://github.com/pytorch/pytorch/issues/71964

frederikwilde commented:
Hi, I just stumbled upon this. I'm not sure if this helps, but I implemented the JVP rule for the complex-valued SVD here: https://github.com/google/jax/pull/5225 and we showed how the derivation works in our paper (https://arxiv.org/pdf/2209.14328.pdf) in appendix D.

lezcano commented:
I have a half-written paper where I discuss how to formalise all this stuff with principal bundles and all that, but I never found the time to finish writing it really :(

frederikwilde commented:
Nice. I would be curious to see it when it's done.
